Introduce specification for OTel Profiles (#4197)
open-telemetry/opentelemetry-specification: specification/profiles/mappings.md

Reviewer: @florianl Please make sure to reference this new directory and file at https://github.com/open-telemetry/opentelemetry-specification/tree/main/specification#contents.

cc @chalin in case you'd need us to add any Hugo-specific things beyond that so it makes it into https://opentelemetry.io/docs/specs/otel/ as expected.

Author: Added Profiles to ToC with https://github.com/open-telemetry/opentelemetry-specification/pull/4197/commits/cf0d197a6324e3cea54cc8bdaee3b2346457124c

Reviewer 2: Thanks for CC'ing me @arminru.

**Per #2558, every section should have a README** -- so make sure that you add a `README.md` file, even if it only contains a title.

For an _example_ of what will be required, see https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/sdk_exporters/README.md?plain=1

Author: Added a README as requested with https://github.com/open-telemetry/opentelemetry-specification/pull/4197/commits/a3630f10b8611c861fefe9cf3fbc847bcb5f7c3c

# Mappings

**Status**: [Development](../document-status.md), except where otherwise specified

> This document defines how to record `Mapping`s and their required attributes.

Reviewer: I think we should introduce what a Mapping is if we build on that here in the spec.

Alternatively, we could at least link to https://github.com/open-telemetry/oteps/blob/main/text/profiles/0239-profiles-data-model.md#message-mapping to give some minimal context.

Author: Added link to oteps wit hhttps://github.com/open-telemetry/opentelemetry-specification/pull/4197/commits/cf398674924fb46f46bb1ced1df9dbd35a165f92.

<!-- toc -->

- [Attributes](#attributes)

<!-- tocstop -->

## Attributes

A message representing a `Mapping` MUST have at least one of the following
attributes:

- `process.executable.build_id.go`

> - `process.executable.build_id.gnu`

Reviewer: Please add a link to the respective semconv definition once https://github.com/open-telemetry/opentelemetry-specification/pull/4197 is merged (as a follow-up as the PRs would otherwise form a circular dependency 😄).

Author: Sure - https://github.com/open-telemetry/semantic-conventions/pull/1329 and https://github.com/open-telemetry/opentelemetry-specification/pull/4197 kind of depend on each other.

Author: Added a link to process attributes with https://github.com/open-telemetry/opentelemetry-specification/pull/4197/commits/05d77599e058e7a6d49f2891ce8562b6fd58432b.

# Mappings

**Status**: [Development](../document-status.md), except where otherwise specified

This document defines the required attributes of [`Mapping`](https://github.com/open-telemetry/oteps/blob/main/text/profiles/0239-profiles-data-model.md#message-mapping)s.

<!-- toc -->

- [Attributes](#attributes)
  * [Algorithm for `process.executable.build_id.profiling`](#algorithm-for-processexecutablebuild_idprofiling)

<!-- tocstop -->

## Attributes

A message representing a `Mapping` MUST have at least one of the following
attributes:

- `process.executable.build_id.gnu`
- `process.executable.build_id.go`
- `process.executable.build_id.profiling`

> If possible all the above listed attributes SHOULD be present in a `Mapping`.

Reviewer 3: ```suggestion
If possible all the above listed attributes SHOULD be present in a `Mapping`. To promote interoperability, it is RECOMMENDED for `process.executable.build_id.profiling` to be present in every `Mapping`.
```

### Algorithm for `process.executable.build_id.profiling`

In some environments GNU and/or Go build_id values are stripped or not usable - for example Alpine

> Linux which is often used as a base for Docker environments. In such cases, profiling generates
> build ids using a specified algorithm. The algorithm hashes the first and last page of a file
> together with its length:

Reviewer 3: ```suggestion
Linux which is often used as a base for Docker environments. In such cases, profiling generates
build ids using a specified algorithm. For that reason and to promote interoperability, a deterministic build_id generation algorithm that hashes the first and last page of a file
together with its length is defined as:
```

> SHA256(first4k, last4k, fileLen)

Reviewer 3: ```suggestion
TRUNC(SHA256(first4k, last4k, fileLen))

where `TRUNC` returns the first 16bytes of the input, `fileLen` is the 8byte big-endian serialization of the file length and `first4k`, `last4k` are the first and last 4096 bytes of the file (may overlap).
```

Should we also define the format of the build id? (e.g. hex string or base64)? Maybe we should also give an example input and the corresponding output to avoid ambiguity?

> TRUNC(SHA256(first4k, last4k, fileLen))

Reviewer 3: Should we also define the format of the build id? (e.g. hex string or base64)? Maybe we should also give an example input and the corresponding output to avoid ambiguity?

Reviewer 4: You should call out that you're defining the format within semconv: https://github.com/open-telemetry/semantic-conventions/pull/1329

I'd have both documents reference each other if possible, or just call out the format will be owned in semconv.

Reviewer 3: Right, that was the intention but we left it out as this PR isn't merged yet. We'll update semconv once we have a specification to link to.

> TRUNC(SHA256(first4k, last4k, fileLen))

Reviewer 3: Switching to the pseudocode we have in the [repository](https://github.com/open-telemetry/opentelemetry-ebpf-profiler) `README`, for better clarity:

````suggestion
```
Input   ← Concat(File[:4096], File[-4096:], BigEndianUInt64(Len(File)))
Digest  ← SHA256(Input)
BuildID ← Digest[:16]
```
````

> where `TRUNC` returns the first 16bytes of the input, `fileLen` is the 8byte big-endian serialization of the file length and `first4k`, `last4k` are the first and last 4096 bytes of the file (may overlap).

Reviewer 3: ```suggestion
where `TRUNC` returns the first 16bytes of the resulting digest, input to `SHA256` is passed left to right, `first4k`, `last4k` are the first and last 4096 bytes of the file (may overlap) and `fileLen` is the 8byte big-endian serialization of the file length.
```

> where `TRUNC` returns the first 16bytes of the resulting digest, input to `SHA256` is passed left to right, `first4k`, `last4k` are the first and last 4096 bytes of the file (may overlap) and `fileLen` is the 8byte big-endian serialization of the file length.

Reviewer 5: Are first4k, last4k padded if the file size is less than 4k or is the file content takes as is?

Reviewer 3: Not padded, I'll update the text.

> where `TRUNC` returns the first 16bytes of the resulting digest, input to `SHA256` is passed left to right, `first4k`, `last4k` are the first and last 4096 bytes of the file (may overlap) and `fileLen` is the 8byte big-endian serialization of the file length.

Reviewer 3: ```suggestion
where `Input` is the concatenation of the first and last 4096 bytes of the file (may overlap, not padded) and the 8 byte big-endian serialization of the file length. The resulting `BuildID` the truncation of the hash digest to 16 bytes (128 bits) in hex string form.
```

# Mappings

> **Status**: [Development](../document-status.md), except where otherwise specified

Reviewer 6: ```suggestion
**Status**: [Development](../document-status.md)
```

This document defines the required attributes of [`Mapping`](https://github.com/open-telemetry/oteps/blob/main/text/profiles/0239-profiles-data-model.md#message-mapping) messages.

<!-- toc -->

- [Attributes](#attributes)
  * [Algorithm for `process.executable.build_id.htlhash`](#algorithm-for-processexecutablebuild_idhtlhash)

<!-- tocstop -->

## Attributes

A message representing a `Mapping` MUST have at least one of the following
[process attributes](https://opentelemetry.io/docs/specs/semconv/attributes-registry/process/#process-attributes):

- `process.executable.build_id.gnu`

> - `process.executable.build_id.go`

Reviewer 7: We should probably add some wording to discourage the use of go build id's and link to this comment as an explanation:

https://github.com/golang/go/issues/68652#issuecomment-2274452424

I don't think this needs to hold up this patch, so I'm approving it now.

Author: Added the link to the comment with https://github.com/open-telemetry/opentelemetry-specification/pull/4197/commits/bbf7543243aeb9ab52136bd2d38f57e1dc0a58d1

> - `process.executable.build_id.profiling`

Reviewer 7: Not sure about the name `profiling` for this. How about `htlhash` (`htl` = `head tail length`) or something that is more descriptive? This way we'd keep the door open for [more hash algorithms](https://cloud-native.slack.com/archives/C03J794L0BV/p1722421785163869) in the future.

We're not planning to propose a different algorithm for now, but we'd like to have the option.

Author: sorry for the delay in response :pray: I was away for some days.

Pushed the changes with the suggested hash name.
